### PR TITLE
Daother 8061

### PR DESCRIPTION
**Pull request by j-s-135**
## Summary of Changes:
https://rcsbpdb.atlassian.net/browse/DAOTHER-xxxx

- Implemented client context
- Rewrote shell scripts
- Rewrote JWT disable
- Modularized config file path

## Screenshot of Changes (or test)
(where applicable)

### Checklist:

- [y ] I followed the code style guidelines
- [y ] I have commented my code, particularly in hard-to-understand areas
- [y ] I have tested the code / Added unit test(s)

### CODE REVIEW

**Review by piehld — CHANGES_REQUESTED** [review]
Thanks for the PR @j-s-135. I'm still going through the changes, but the first thing I see that I think should be changed is the moving of the ClientUtils into the 'python' sub-directory. That's not really a standard way of packaging python projects, since the user should  assume the python project is all python. That's kind of why I originally suggested just moving the front-end related code to its own directory, but leave the core parts of the package in the standard directory tree structure (i.e., `rcsb.app.client.ClientUtils`).

Can you please make that directory adjustment? (Apologies I know that affects all the imports too.)

File: rcsb/app/client/python/ClientUtils.py [hunk 1]
```diff
@@ -297,31 +353,31 @@ def download(
...
-            logger.error("Download folder does not exist")
-            return None
         convertedMilestone = None
+        if not milestone or milestone.lower() == "none":
+            milestone = ""
         if milestone and milestone.lower() != "none":
             convertedMilestone = f"-{milestone}"
         else:
             convertedMilestone = ""
         convertedContentFormat = self.fileFormatExtensionD[contentFormat]
```
Comment (piehld) [inline rcsb/app/client/python/ClientUtils.py:369]: Can you explain what the purpose of this is? Why must you convert the milestone? What exactly are you doing to it?

File: rcsb/app/tests-client/testClientUtils.py [hunk 2]
```diff
@@ -241,13 +241,54 @@ def testListDir(self):
...
+        milestone = None
+        partNumber = 1
+        contentFormat = "pdbx"
+        version = 1
+        filename = self.__cU.getFilePathLocal(repoType,depId,contentType,milestone,partNumber,contentFormat,version)
+        print(f'temp file path {filename}')
+        self.assertTrue(os.path.exists(filename), f'error - {filename} does not exist')
+        with open(filename, "rb") as r:
+            bytes = r.read()
+            print(bytes)
```
Comment (piehld) [inline rcsb/app/tests-client/testClientUtils.py:273]: Is this part really necessary for purposes of the test? Might be worth removing if not...

File: setup.py [hunk 3]
```diff
@@ -42,9 +42,9 @@
         "Intended Audience :: Developers",
         "Natural Language :: English",
         "License :: OSI Approved :: Apache Software License",
-        "Programming Language :: Python",
-        "Programming Language :: Python :: 3",
-        "Programming Language :: Python :: 3.9",
+        "Programming Language :: py",
+        "Programming Language :: py :: 3",
+        "Programming Language :: py :: 3.9",
```
Comment (piehld) [inline setup.py:47]: Was this change intentional? Part of me suspects it might have been an unintentional effect of a "Find all and replace" command...

**Review by piehld — CHANGES_REQUESTED** [review]
Thanks again @j-s-135 for all the code updates. I had a couple questions and comments throughout, along with my [comment from earlier](https://github.com/rcsb/py-rcsb_app_file/pull/23#pullrequestreview-1332856545) about the directory structure of the client files. I'd appreciate if you're able to follow-up and/or help elaborate on them.

**Comment by j-s-135** [comment]
1. python subdirectory - the project can't be all python since it's a web project that needs an HTML front end (as POC for the back end, which might or might not change the way we write the backend functions). So I have written some HTML files to be used for testing, and after Weslley said their front end was in jQuery, I wrote another one, so to avoid clutter I thought we should make two separate directories, one for Python and one for HTML. I would also argue that the HTML front end is more important than the Python front end since the Python front end cannot be used from a website.
2. Yes, maybe you're right about overwriting environment vars, but do you have any idea what we would be overwriting? Why wouldn't setting the config_file var to point to the config_file within the project be sufficient?
3. Yes, ConfigProvider would be an even better way to set the config_file var, but what I've done is already an improvement over the previous file-by-file approach. I agree we should improve it further.
4. The "none" string is just a permissive approach that allows the default milestone to be None, "", "None", or "none". It's maybe another front-end issue. From the front end of a web form, the user may get to type in a value for milestone. Similarly, a command-line parameter to a Python front-end is the same way. Since we don't know what sort of front-end we're using, I recommend a permissive approach for now to make front-end testing easier.
5. Sorry, I thought you understood from last week that I couldn't use the client.yml file. It threw an error (try it yourself). I was trying to explain that the client files DO need to have the repository paths. We could put them in the client file, but then the only difference from the config file is the database constants.
6. The milestone that a user inputs or selects from the front end does not have a dash in front of it, but the path string that is formed has a dash before the milestone, so I'm just building the file path strings ahead of time there.
7. I think you're right, we don't need to print the file as part of the unit test. I was just testing it as a potential use case.
8. Again, ConfigProvider would be better, but I haven't gotten to implementing that yet. It's better than what the repo had.
9. I haven't intentionally changed tox.ini or setup.py. I will remove those except for one of the redundant lines.

**Comment by piehld** [comment]
> 1. python subdirectory - the project can't be all python since it's a web project that needs an HTML front end (as POC for the back end, which might or might not change the way we write the backend functions). So I have written some HTML files to be used for testing, and after Weslley said their front end was in jQuery, I wrote another one, so to avoid clutter I thought we should make two separate directories, one for Python and one for HTML. I would also argue that the HTML front end is more important than the Python front end since the Python front end cannot be used from a website.

I see your points, but this kind of comes back to the idea that this project is intended to be re-applicable to different use cases, not targeted for specifically the DepUI use case. For sure it's valuable to have the different use-case code and examples in the repo, but it should be clearly separated from the core components that are meant to be more general. It also kind of comes down to what users of this package are going to use—are they actually going to import the code under 'html'? My understanding is not directly, but rather that Weslley will use that code as a starting point. However, users are going to import the ClientUtils code, and since they are working in python and the package is uploaded to PyPI it just seems a little redundant to add 'python' to the module path. If there are examples of doing this in any other pypi projects though, I may be over-assuming what I think "standard packaging" is. 

> 2. Yes, maybe you're right about overwriting environment vars, but do you have any idea what we would be overwriting? Why wouldn't setting the config_file var to point to the config_file within the project be sufficient?

What if the user wants to tell the project that they want to use a config file somewhere else on their system? They should ideally be able to set the `CONFIG_FILE` environment variable on their host machine and then boot up the API service, and then the service would use that pre-defined variable. But the way it's working now, it'll overwrite the user's variable with the config file in the project, which could cause a lot of damage if they weren't aware of that overwrite. It would be a nice feature if the user could control what config file is used by way of presetting an environment variable.

> 3. Yes, ConfigProvider would be an even better way to set the config_file var, but what I've done is already an improvement over the previous file-by-file approach. I agree we should improve it further.

OK, sure no problem. We should keep note of this then for moving forward. Was just wondering why the separation.

> 4. The "none" string is just a permissive approach that allows the default milestone to be None, "", "None", or "none". It's maybe another front-end issue. From the front end of a web form, the user may get to type in a value for milestone. Similarly, a command-line parameter to a Python front-end is the same way. Since we don't know what sort of front-end we're using, I recommend a permissive approach for now to make front-end testing easier.

OK thanks for explaining, that seems reasonable.

> 5. Sorry, I thought you understood from last week that I couldn't use the client.yml file. It threw an error (try it yourself). I was trying to explain that the client files DO need to have the repository paths. We could put them in the client file, but then the only difference from the config file is the database constants.

OK I understand, but the failures you were facing were just for the `testClientUtils.py` tests, right? That should be fixable, but I suppose this could be made into a separate ticket anyway, along with the test-config.yml idea.

> 6. The milestone that a user inputs or selects from the front end does not have a dash in front of it, but the path string that is formed has a dash before the milestone, so I'm just building the file path strings ahead of time there.

Huh OK, I guess that works, although I would have leaned towards building the file path string in one swoop and putting the dash in at that point, rather than modifying the variable to have a dash in it itself. But I see you're using a different variable name for that, so no harm done that way anyway.

> 7. I think you're right, we don't need to print the file as part of the unit test. I was just testing it as a potential use case.
OK sounds good.

> 8. Again, ConfigProvider would be better, but I haven't gotten to implementing that yet. It's better than what the repo had.
Sure understand.

> 9. I haven't intentionally changed tox.ini or setup.py. I will remove those except for one of the redundant lines.
OK great, thanks.

**Comment by j-s-135** [comment]
That makes sense about imports. We could modularize the javascript files so that people could import them, but it wouldn't be through a Python package. So the question then is where to put the HTML files?

**Comment by piehld** [comment]
I think it would be cleanest to create a directory called `fe` or `front-end`, into which you put the `html` directory along with all the front-end related python files. You could keep that front-end directory in the `rcsb/app/client` directory if you think that makes the most sense for them. Then I would move the `ClientUtils.py` file back to under `rcsb/app/client` directly (along with the `__init__.py` file).

**Comment by j-s-135** [comment]
You're suggesting guy, client, and all the HTML files in the front-end directory within the client directory and ClientUtils as a child of the client directory?

**Comment by piehld** [comment]
Yeah, this is what I had in mind:
```
rcsb/app/client/
  __init__.py
  ClientUtils.py
  front-end/
    html/*
    gui.py
    onedep_logo.png
    client.py
```

**Comment by j-s-135** [comment]
All right. Then front-end needs an init for gui and client.

File: setup.py [hunk 4]
```diff
@@ -42,8 +42,6 @@
         "Intended Audience :: Developers",
         "Natural Language :: English",
         "License :: OSI Approved :: Apache Software License",
-        "Programming Language :: py",
-        "Programming Language :: py :: 3",
         "Programming Language :: py :: 3.9",
```
Comment (piehld) [inline setup.py:45]: Interesting, do we not need these classifiers at all? Are they purely redundant, or is there any value in keeping them here? I'm not entirely sure how these are used...

Comment (j-s-135) [reply setup.py:45]: Sorry, didn't fix that one. I will make another update.

File: rcsb/app/client/ClientUtils.py [hunk 5]
```diff
@@ -417,3 +503,62 @@ def listDir(self, repoType: str, depId: str) -> list:
...
+                      unit_test:bool = False,
+                      wfInstanceId: str = None,
+                      sessionDir:str = None
+                      ):
+            downloadFolder = None
+            allowOverwrite = True
+            returnTempFile = True
+            deleteTempFile = False
+            file = self.download(repoType, depId, contentType, milestone, partNumber, contentFormat, version, hashType, downloadFolder, allowOverwrite, returnTempFile, deleteTempFile)
+            return file.name
```
Comment (piehld) [inline rcsb/app/client/ClientUtils.py:557]: @j-s-135 I think the indentation here is off

File: rcsb/app/client/ClientUtils.py [hunk 6]
```diff
@@ -417,3 +503,62 @@ def listDir(self, repoType: str, depId: str) -> list:
...
+
+    def dir_exist(self, repositoryType, depId):
```
Comment (piehld) [inline rcsb/app/client/ClientUtils.py:559]: Can you change this to `dirExist`? For consistency with all the other methods

**Review by piehld — APPROVED** [review]
Thanks @j-s-135!

**Pull request merged**